📜Create mouse tutorial (#53)
KinsonDigital/VelaptorDocs: docs/tutorials/guides/mouse-input.mdx

---
sidebar_position: 7
---

import { TutorialLink } from "@site/src/components/TutorialLink";

# Mouse Input

> :::note Project Templates

Reviewer: This syntax doesn't look good in rich diff.  It's located throughout the whole document.
```suggestion
:::note Project Templates
```

> Go to the <TutorialLink projectName="MouseInput"/> tutorial project to see a fully working example.

Reviewer: Link broken 
```suggestion
Go to the <TutorialLink projectName="MouseInput"/> tutorial project to see a fully working example.
```

In this tutorial, you will learn how to get the state of the mouse so you can make decisions on that state.  This will open
the doors giving your game the control that it needs outside of using a keyboard.

:::note project example
Go to the <TutorialLink projectName="KeyboardInput"/> tutorial project to see a fully working example.
:::

## <span class="color-step">Step 1: Project setup</span>

Set up an empty Velaptor project.

:::note REFER TO
For more info, refer to the [Project Setup](./project-setup) guide.
:::

The only game loop method required for this tutorial will be the `OnUpdate()` method, which we
will be used to update the game state and resize the window.

The rest of the game loop methods described in the _**project setup guide**_ can be removed.

## <span class="color-step">Step 2: Creating game state</span>

Let's create some class fields to help us control the window size.

### <span class="color-sub-step">Step 2.1: Creating the class fields</span>

Create some class fields that can hold the state of the mouse and also the amount of how much we want to change

> the height of the window.

Reviewer: Reword
```suggestion
Create some class fields that can hold the state of the mouse and window height changes.
```

Author: What you requesting does not make sense to what the class fields are and what they do.

Just like it says, one of the class fields holds the **AMOUNT** that the window height will change when the change is applied.

Recheck what I did and see if that is good to go.

Use the factory to create the mouse object in the `Game` constructor.

```csharp
public Game() => this.mouse = InputFactory.CreateMouse();
```

## <span class="color-step">Step 3: Changing the window size</span>

Perform the following steps to change the size of the window by using the mouse.

### <span class="color-sub-step">Step 3.1: Getting the mouse state</span>

We can get the state of the mouse by using the `mouse` object that we saved to the class field.  In the `OnUpdate()` method,
use the `mouse` object to get the current state of the mouse.  Also, save the current mouse state as the previous mouse state.
This way, when the next frame is executed, we can compare the current mouse state to the previous mouse state to see 
if there are differences between the two states to detect if a button has been pressed or the mouse has moved.

```csharp
protected override void OnUpdate(FrameTime frameTime)
{
	var currMouseState = this.mouse.GetState();

	this.prevMouseState = currMouseState;

	base.OnUpdate(frameTime);
}
```

### <span class="color-sub-step">Step 3.2: Changing the window width</span>

> Next, between the two mouse states in the `OnUpdate()` method, add the code below to change the width of the window.

Reviewer: Reword
```suggestion
Next, between the two mouse states in the `OnUpdate()` method, change the width of the window by adding the code below:
```

```csharp
private void OnUpdate(FrameTime frameTime)
{
	... 

	// If the left mouse button is currently being held in the down position
	if (currMouseState.GetButtonState(MouseButton.LeftButton))
	{
		// The position of the mouse for the current frame
		var currMousePos = currMouseState.GetPosition();

		// The position of the mouse in the previous frame
		var prevMousePos = this.prevMouseState.GetPosition();

		// Add the mouse movement amount to the current width of the window
		Width += (uint)(currMousePos.X - prevMousePos.X);
	}

	this.prevMouseState = currMouseState;

	base.OnUpdate(frameTime);
}
```

> This code checks if the left mouse button is currently being held down. If it is, then the width of the window will be
> changed based on the difference between the current and previous mouse positions.  As you move the mouse left and right,
> it will take the difference between the position on the X-axis and apply that to the width of the window.

Reviewer: Reword
```suggestion
This code checks whether  or not the left mouse button is currently being held down. If it is, then the width of the window will change based on the difference between the current and previous mouse positions.  As you move the mouse left and right, it will take the difference between the position on the X-axis and apply that to the width of the window.
```

The important part here is the `this.prevMouseState = currMouseState`.  As long as the logic is added between the current
and previous mouse states, there is something to compare to know if the mouse has moved.

### <span class="color-sub-step">Step 3.3: Changing the window height</span>

Now we can add the code to change the height of the window as we scroll the mouse wheel.

> Add the following code right below the `if` block added in the previous step.

Reviewer: Reword
```suggestion
Below the `if` block we added in the previous step, add the code below: 
```

```csharp
private void OnUpdate(FrameTime frameTime)
{
	... 

	// Get the current value of the mouse scroll wheel
	var wheelValue = currMouseState.GetScrollWheelValue();

	// If the mouse wheel was not scrolled, exit the method
	if (wheelValue != 0)
	{
		// Check if the mouse wheel was scrolled down or up.
		// If the value is negative, the mouse wheel has been scrolled in the down direction.
		var scrolledDown = wheelValue < 0;

		// Change the height of the window based on the direction the mouse wheel was scrolled.
		Height = scrolledDown
			? Height - HeightChangeAmount
			: Height + HeightChangeAmount;
	}

	this.prevMouseState = currMouseState;

	base.OnUpdate(frameTime);
}
```

> This code will get the current value of the mouse wheel and use this to detect which direction the mouse wheel was scrolled.
> When the mouse wheel is scrolled down toward you, it returns a negative value and when it is scrolled up away from you,
> it returns a positive value.

Reviewer: Reword

```suggestion
This code gets the current value of the mouse wheel and uses it to detect which direction the mouse wheel was scrolled.
When the mouse wheel is scrolled down toward you, it returns a negative value and when it is scrolled up away from you, it returns a positive value.
```

If the mouse wheel has not been scrolled at all for the current frame, it will be 0.  Only change the height if the mouse
has been scrolled.

If the mouse was scrolled in either direction, increase or decrease the height of the window.
That's it!!  Now we are controlling the width and height of the window using the mouse!!
This of course can be used for _**ANYTHING**_.  Character movement, aiming a weapon, and game menu item selection
some examples of what the mouse can be used for in a game.

### <span class="color-sub-step">Step 3.4: Run it</span>

Run the application and you should be able to change the width and height of the window using the mouse.
To change the width of the window, hold down the left mouse button and move the mouse from left to right.
To change the height of the window, scroll the mouse wheel up or down.

## <span class="color-step">Step 4: Bonus</span>

> It would be nice to see the current size of the window. To do this, we can override the `OnResize()` method

Reviewer: Up to you if you wanna reword this
```suggestion
It would be nice to see the current size of the window, right? To do that, we can override the `OnResize()` method
```
